Updated problem 2 for week 1 - Paritosh Sharma

pari08tosh opened the pull request:
Submitting the 10th update to problem 1.2 according to suggestions in previous PR. Have made the following changes:-
1. Improved variable names and type to reduce confusion.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+];
+
+var parsed_json = {};
+var err = [],flag,j;
```
Spaces after `,`

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+
+// add new argument entries here
+
+var command_args  = [
```
You have an extra space before `=`.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+      required: false,
+      position: null
+   }
+];
```
Is this the best way to provide input? How about the builder pattern?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+
+function print_json(err){
+   if(err)
+   {
```
You are inconsistently using these braces. In case of functions, you put them on the same line. In case of `if` and the next `for`, you put them on the next line. Please fix that. I prefer same line. Also, space before `{`

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+      console.log('ERROR: ' + err);
+      return;
+   }
+   for(i=0;i<command_args.length;i++)
```
Proper spacing:
```
for (i = 0; i < command_args.length; i++) {
```

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+   for(i=0;i<command_args.length;i++)
+   {
+      if(command_args[i].required === true && !(command_args[i].key in parsed_json)){
+         console.log( "ERROR: Value of " + command_args[i].arg + " required");
```
If this function is called "print_json", it is not allowed to do validation. Please move this code out.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+   return;
+}
+
+arg_parser(entered_args, print_json);
```
I had to search for where `entered_args` is defined. Which meant that my flow of reading was broken. Let's try and avoid that. Move all statements outside functions to the bottom if possible.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+   for(var i=0; i<entered_args.length; i++){
+      var str = entered_args[i];
+      flag = 0;
+      if(str.search('-') === -1){
```
Why shouldn't it appear inside the string? Like `init-clean`

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+
+arg_parser(entered_args, print_json);
+
+function arg_parser(entered_args, callback){
```
Most of your variable / function names don't meet the conventions mentioned in the style guide.
Also, why not just return the result. Why do you need a callback? This isn't async.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,127 @@
+         {
+            if(command_args[j].arg === str[0]){
+               flag=1;
+               if(command_args[j].type === 'String'){
```
Avoid raw strings like this. Prefer enums or constants, which will result in an error if mistyped.

pari08tosh commented:
I have updated the solution to the problem according to your suggestions. Please review it again.

pari08tosh commented:
Updated problem 1.2 according to the changes requested in the previous pull request. Have made the following updates -

    1. Updated input format to a more user friendly one.
    2. Edited and updated code according to the style guide.
    3 . Removed Bugs

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+   return;
+}
+
+// add new arguments here in the format argBuilder('key', 'argumentName', 'type', 'required(Boolean)', position);
```
This separation of core logic and input/output is so much cleaner :)

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+argBuilder('name', '--name', 'String', false, null);
+argBuilder('verbrose', '-v', 'Boolean', false, null);
+argBuilder('local', '--local', 'Boolean', false, null);
+argBuilder('remote', '--remote', 'Boolean', false, null);
```
Next challenge: Notice how I'm forced to depend on your comment to understand the meanings of the last 2 arguments. It isn't intuitive enough from the code itself. This is why, in general, boolean arguments are discouraged. Can you modify this builder so that the meanings of all inputs are obvious?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+argBuilder('remote', '--remote', 'Boolean', false, null);
+
+var parsedJson = {};
+var err = [], flag, j;
```
I know this works, but it isn't intuitive how you declare a variable here, but use it before. Unfortunately, while `let` and `const` are slightly better, they have the same behavior in this regard. Why not just make these local variables.

In general, you want to minimize the amount of global state.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+      console.log('ERROR: ' + err);
+      return;
+   }
+   for(i = 0; i < argList.length; i++) {
```
Space after `if`, `for`, etc. Applies everywhere.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+      else {
+         str = str.split('=');
+         for(j = 0; j < argList.length; j++)
+         {
```
You're starting the block from the next line. But in other places (like line 38), you're starting it from the same line. You need to be consistent.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+
+function argParser(inputArgs) {
+
+   for(var i = 0; i < inputArgs.length; i++) {
```
Tip: This isn't something you'll find in a style guide, but lots of people do this. Use `ii` instead of `i`, `jj` instead of `j`, etc. This makes those variables much easier to find. And if you use multiple cursors, you will immediately see the benefits.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+};
+
+function argBuilder(key, arg, type, required, position) {
+   if(key === null){
```
Space before `{`. Applies everywhere.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+                  else {
+                     parsedJson[argList[j].key] = str[1];
+                  }
+               }
```
This section could be improved. Note how your validation is separate from your type definition. This means that if I want to add a new type, I'll need to update 2 places which is inconvenient. You want to simplify it to 1 place.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+                  }
+               }
+               else if(argTypeEnum[argList[j].type] === 3) {
+                  if( isNaN(str[1]) || str[1] * 1 < 0) {
```
Whitespace isn't right. Should be `if (isNan`

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+                     console.log(`ERROR: Invalid value of argument '${str[0]}'. Must be a positive integer`);
+                     return;
+                  }
+                  else {
```
While consistency isn't a problem in this case, how about `} else {`?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+
+   // validating. Add aditional validations here
+
+   if(err.length > 0) {
```
`err` isn't set anywhere, as far as I see.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,121 @@
+var err = [], flag, j;
+var inputArgs = process.argv.slice(2);
+
+argParser(inputArgs);
```
Console log the result, instead of print from inside the method. That can make the method unused for other purposes.

pari08tosh commented:
Third update of problem 1.2. Modified code according to changes requested in last PR. Have made the following changes :-
1. Modified input method to improve user friendliness.
2. Modified method to add new argument types.
3. Simplified main parser method.
4. Restructured functions to improve code readability and usability.
5. Removed style errors and other bugs.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+   'String': 'true',
+   'Boolean': 'true',
+   'Int': '!isNaN(argValue)',
+   '+Int': '!isNaN(argValue) && argValue * 1 >= 0' // represents + integers.
```
Interesting. This works, but what if you use functions instead? `eval`ing string is generally not a good idea, due to performance/security issues.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+argBuilder('name', '--name', 'String', 'notRequired');
+argBuilder('verbrose', '-v', 'Boolean', 'notRequired');
+argBuilder('local', '--local', 'Boolean', 'notRequired');
+argBuilder('remote', '--remote', 'Boolean', 'notRequired');
```
Much better. You got read of all the booleans, but we have different problems with strings. Basically, it is really easy to make typing mistakes that no-one notices, and so you might type in `notReqired` and your code does unexpected things. As a result, it is better to use Enums.
```
const IS_REQUIRED = {
  YES: true,
  NO: false,
};
```
This makes things clearer, and safe too. Another approach is:
```
argBuilder({
  name: 'key',
  isRequired: true,
  ...
})
```

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+      }
+   }
+   if (printCheck === true)
+      printParsedArgs();
```
As mentioned in the post yesterday, a method called `argValidator` has no business print stuff. Remove this.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+//argument validator function. Add additional validations here.
+//Pass printCheck = true to print parsed arguments.
+
+function argValidator(printCheck) {
```
Since all you do is to ensure that required arguments are prevent, why not call it that. This name makes it sound like you're going to validate the individual values.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+function argParser(inputArgs) {
+   var  ii, jj, flag, argValue;
+   for (ii = 0; ii < inputArgs.length; ii++) {
+      var str = inputArgs[ii];
```
`str` tells me what type it is, but not what it contains. Better name?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+   for (ii = 0; ii < inputArgs.length; ii++) {
+      var str = inputArgs[ii];
+      flag = 0;
+      if (str.search('-') !== 0) {
```
`indexOf` is clearer, with the same results.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+// add new arguments here in the format argBuilder('key', 'argumentName', 'type', 'required/notRequired', position(if necessary));
+
+argBuilder('command', null, 'String', 'notRequired', 0);
+argBuilder('subcommand', null, 'String', 'notRequired', 1);
```
The position argument is still not obvious, unfortunately.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+      if (str.search('-') !== 0) {
+         for (jj = 0; jj < argList.length; jj++) {
+            if (argList[jj].position === ii) {
+               flag = 1;
```
Instead of using a flag, you could just continue after the next line. And show a common error message at the bottom.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+            if (argList[jj].arg === str[0]) {
+               flag = 1;
+               argValue = str[1];
+               if (eval(argTypeConditions[argList[jj].type]))  {
```
As mentioned above, prefer not use `eval`.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,112 @@
+                  if (argList[jj].type === 'Boolean')
+                     parsedJson[argList[jj].key] = true;
+                  else
+                     parsedJson[argList[jj].key] = argValue;
```
So even an Int will be stored as a string? That's not elegant. How about your validator function returns the value, cast in the proper type. That way, you dont have to special case Boolean.

pari08tosh commented:
Fourth update of problem 1.2. Modified code according to changes requested in last PR. Have made the following changes :-
1. Modified input method to improve user friendliness.
2. Modified method to add new argument types.
3. Simplified main parser method.
4. Restructured functions to improve code readability and usability.
5. Improved data type handling and error handling.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,117 @@
+
+var parsedObj = argParser(inputArgs);
+
+if (parsedObj instanceof Error) {
```
You shouldn't be returning Error objects. You should throw them, and use try-catch.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,117 @@
+// Enter 'true' if no conditions
+
+argTypeConditions = {
+  'String': [function(argValue) { return true; }, 'String'],
```
Prefer objects instead of arrays like this, since it gives you an opportunity to specify intent using keys.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,117 @@
+function argBuilder(argObj) {
+  if (argObj.key === null) {
+    console.log(`Error: Key of argument '${arg}' cannot be null`);
+    process.exit(1);
```
Throw errors. Don't exit early like this.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,117 @@
+
+  for (i = 0; i < argList.length; i++) {
+    if (argList[i].isRequired === true && !(argList[i].key in tempObj)) {
+      return new Error(`Value of '${argList[i].arg}'  required`);
```
throw new Error ...

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,117 @@
+  return true;
+}
+
+function printParsedArgs(parsedObj) {
```
Eliminate this. Use console.log directly.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,108 @@
+
+function argGroupBuilder(argObj) {
+  var ii;
+  for(ii = 0; ii < argObj.argList.length; ii++) {
```
According to style guide, there needs to be a space after `for`.
Also, not sure why you didn't just do `var ii = 0;`, instead of declaring it separately.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,108 @@
+argGroupBuilder({ groupId: 3, isRequired: true, argList: [ { key: 'key', arg: '--key', type: '+Int', position: null } ] });
+argGroupBuilder({ groupId: 4, isRequired: false, argList: [ { key: 'name', arg: '--name', type: 'String', position: null } ] });
+argGroupBuilder({ groupId: 5, isRequired: false, argList: [ { key: 'verbrose', arg: '-v', type: 'Boolean', position: null } ] });
+argGroupBuilder({ groupId: 6, isRequired: false, argList: [ { key: 'local', arg: '--local', type: 'Boolean', position: null }, { key: 'remote', arg: '--remote', type: 'Boolean', position: null } ] });
```
Uh ... this API is rather inconvenient to use. The user has to provide these complex objects with multiple keys, and they wouldn't know if they made any mistakes until they actually run your `argParser` method, and maybe not even then. There is a "fail fast" philosophy, suggesting that you want errors to be generated ASAP.

At this point, I'd suggest looking at the API provided by https://www.npmjs.com/package/commander (or other examples) to see how simple they are to use, and compare what you have with them. After that, can you improve your API?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,108 @@
+return true;
+}
+
+function argParser(inputArgs) {
```
This is a better name for a class, not a static function.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,108 @@
+      throw new Error(`Multiple arguments of group same as argument '${groupList[ii].argList[0].arg}' present.`);
+    }
+  }
+return true;
```
indenting issue.

kaustubh-karkare replied:
This is still a problem.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+  if (!groupName) {
+    throw new Error(`Group name required while creating new group`);
+  }
+  let newGroupObj = {
```
This can be `const`, instead of `let`.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+  }
+  if (!(argObj.type in argTypeFunctions)) {
+    throw new Error(`Invalid type of argument '${argObj.arg}'`);
+  }
```
This sections looks duplicated.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+      if (!(argObj.type in argTypeFunctions)) {
+        throw new Error(`Invalid type of argument '${argObj.arg}'`);
+      }
+      for (let ii = 0; ii < groupList.length; ii++) {
```
Instead of looping through this, why not just add `newGroupObj` here? It should work.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+    throw new Error(`Invalid type of argument '${argObj.arg}'`);
+  }
+  let argGroupObj = { isRequired: argObj.isRequired || false, argList: [ { key: argObj.key, arg: argObj.arg, type: argObj.type, position: argObj.position } ] };
+  groupList.push(argGroupObj);
```
Interesting.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+function parseArg() {
+  let ii, jj, kk, argValue, arg, parseObj = {};
+  outerLoop:
+    for (ii = 0; ii < inputArgs.length; ii++) {
```
Damn, this is dangerous. Your `inputArgs` is defined in the outer scope, but it should be an argument (given you're using it like one on line 141). I disagree with how JavaScript handles this, because of this exact reason.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+        arg = arg.split('=');
+        for (jj = 0; jj < groupList.length; jj++) {
+          for (kk = 0; kk < groupList[jj].argList.length; kk++) {
+            if (groupList[jj].argList[kk].arg === arg[0]) {
```
Shouldn't it be "key" here?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+          }
+        }
+      } else {
+        arg = arg.split('=');
```
So at different point in code, this variable contains different types of things? Given the same name, this is bad naming.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+            if (groupList[jj].argList[kk].arg === arg[0]) {
+              argValue = arg[1];
+              argValue = argTypeFunctions[groupList[jj].argList[kk].type](argValue);
+              if (argValue) {
```
What if the boolean `false` was valid?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,145 @@
+          }
+        }
+      }
+      throw new Error(`Invalid argument '${inputArgs[ii]}' entered. Check Type and Key`);
```
"key" is a term that people who read your code understand. If they haven't, it is confusing.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,147 @@
+    name: groupName,
+    isRequired: false,
+    argList: [],
+    addGroupArg: function(argObj) {
```
For simple functions like this, consider using the arrow syntax.
```
addGroupArg: argObj => addArg(argObj, this),

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,147 @@
+    throw new Error(`Invalid type of argument '${argObj.arg}'`);
+  }
+  if (groupObj) {
+    for (let ii = groupList.length-1; ii >= 0; ii--) {
```
`groupList.forEach(...)` is easier to read, since you dont have to deal with index.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,147 @@
+function parseArg(inputArgs) {
+  let ii, jj, kk, argValue, inputArg, parseObj = {};
+  outerLoop:
+    for (ii = 0; ii < inputArgs.length; ii++) {
```
Yeah, you really want to use `forEach` to avoid the indices, which make things harder to understand.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,154 @@
+    });
+    if (argGroup.isRequired === true) {
+      if (count === 0) {
+        throw new Error(`At least one argument from group of argument '${ argGroup.argList[0].argKey }' required.`);
```
This error message could be confusing if you didn't use any groups.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,154 @@
+            parseObj[arg.argName] = inputArg[0];
+            argParsed = true;
+            return true;
+          } else return false;
```
Always use blocks for all composite statements like if/else/while/for/etc.
```
} else {
  return false;
}
```
Applies below too.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,154 @@
+        return argGroup.argList.some((arg) => {
+          if (arg.position === index) {
+            parseObj[arg.argName] = inputArg[0];
+            argParsed = true;
```
You dont need this.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,154 @@
+      });
+    } else {
+      argParsed = groupList.some((argGroup) => {
+        return argGroup.argList.some((arg) =>{
```
`.some(arg => {`

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,154 @@
+      argParsed = groupList.some((argGroup) => {
+        return argGroup.argList.some((arg) => {
+          if (arg.position === index) {
+            parseObj[arg.argName] = inputArg[0];
```
Use the argTypeFunctions here?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,166 @@
+  }
+
+  if (groupObj) {
+    groupList.some(argGroup => {
```
You're not using the result of the `some` call. What if you dont find the group?

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,166 @@
+        if (argGroup.argList.length === 1) {
+          throw new Error(`Argument '${argGroup.argList[0].argKey}' required`);
+        } else {
+          throw new Error(`At least one argument from group of argument '${ argGroup.argList[0].argKey }' required.`);
```
Much better.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,166 @@
+    }
+
+    if (count > 1) {
+      throw new Error(`Multiple arguments of group same as argument '${ argGroup.argList[0].argKey }' present.`);
```
I'm wondering if the group name is more useful than argKey.

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,166 @@
+
+// main argument parser function
+function parseArg(inputArgs) {
+  let parseObj = {};
```
s/let/const

kaustubh-karkare commented on solutions/paritosh-sharma/week-1/problem-1.2.js:
```diff
@@ -0,0 +1,166 @@
+    let argParsed = false;
+    let argValue = null;
+    let keyLessArg;
+    inputArg = inputArg.split('=');
```
In general, it isn't a good idea to store different types of information in the same variable. You will confused readers about whether this contains an array or a string.

pari08tosh commented:
Submitting the 10th update to problem 1.2 according to suggestions in previous PR. Have made the following changes:-
1. Improved variable names and type to reduce confusion.
